### ISSUE TITLE
libmamba Non-conda folder exists at prefix

### ISSUE THREAD

**Issue by joelive** [jiarong/VirSorter2#222]
When I run the virsorter setup command, it always report error.

(virsort2) bo@bioinofo2:~$ virsorter setup -d virsorter_db/
[2024-10-24 02:47 INFO] VirSorter 2.2.4
[2024-10-24 02:47 INFO] /home/bo/miniconda3/envs/virsort2/bin/virsorter setup -d virsorter_db/
[2024-10-24 02:47 INFO] Setting up VirSorter2 database; this might take ~10 mins and only needs to be done once.
[2024-10-24 02:47 INFO] saving /home/bo/virsorter_db as DBDIR to config file /home/bo/miniconda3/envs/virsort2/lib/python3.10/site-packages/virsorter/template-config.yaml
CreateCondaEnvironmentException:
Could not create conda environment from /home/bo/miniconda3/envs/virsort2/lib/python3.10/site-packages/virsorter/rules/../envs/vs2.yaml:
error    libmamba Non-conda folder exists at prefix
critical libmamba Aborting.

  File "/home/bo/miniconda3/envs/virsort2/lib/python3.10/site-packages/snakemake/deployment/conda.py", line 350, in create
[2024-10-24 02:47 ERROR] CreateCondaEnvironmentException:
Could not create conda environment from /home/bo/miniconda3/envs/virsort2/lib/python3.10/site-packages/virsorter/rules/../envs/vs2.yaml:
error    libmamba Non-conda folder exists at prefix
critical libmamba Aborting.

  File "/home/bo/miniconda3/envs/virsort2/lib/python3.10/site-packages/snakemake/deployment/conda.py", line 350, in create
[2024-10-24 02:47 INFO] First attempt failed; trying the second time.
CreateCondaEnvironmentException:
Could not create conda environment from /home/bo/miniconda3/envs/virsort2/lib/python3.10/site-packages/virsorter/rules/../envs/vs2.yaml:
error    libmamba Non-conda folder exists at prefix
critical libmamba Aborting.

  File "/home/bo/miniconda3/envs/virsort2/lib/python3.10/site-packages/snakemake/deployment/conda.py", line 350, in create
[2024-10-24 02:47 CRITICAL] Command 'snakemake --snakefile /home/bo/miniconda3/envs/virsort2/lib/python3.10/site-packages/virsorter/rules/setup-retry.smk --directory /home/bo/virsorter_db --quiet --config Skip_deps_install=False --jobs 64 --rerun-incomplete --latency-wait 600 --nolock  --use-conda --conda-prefix /home/bo/virsorter_db/conda_envs --conda-frontend mamba ' returned non-zero exit status 1.

**Comment by ChaoLab** [comment]
Hi Jiarong,
I have the same issue as @joelive. I guess it is a conda env setting-up problem.
Can you help to solve this issue?

Best,
Chao

**Comment by jiarong** [comment]
Sorry, I was quite busy last week. 
If you do not have to conda/mamba, I recommend installation [option 3](https://github.com/jiarong/VirSorter2?tab=readme-ov-file#option-3). Or else, try to remove the database directory (`/home/bo/virsorter_db`) and rerun `virsorter setup`. It looks like there was an failed attempt before.

**Comment by handibles** [comment]
Seeing this also on a fresh env. After failed installs, there's a `conda_envs` dir created in the chosen `db` dir, possibly `mamba` is attempting to install the `setup.smk` dependencies alongside the database, which is going awry (i.e. `non-conda folder`). See `line:471` of `virsorter.py`, and possibly `line:155` of `setup.smk`. 

In any case: doing all of the below:
 - A fresh `mamba env`
 - deleting the target `db` dir
 - using `-s`(kip-dependencies) during `virsorter setup` 

allowed setup to complete. Hopefully of use - thanks to the devs for their time.

**Comment by joelive** [comment]
I have solved the problem by mamually mamba install all packages

**Comment by handibles** [comment]
> I have solved the problem by mamually mamba install all packages

@joelive  glad it works! My comment above doesn't allow you to run virsorter, as it still attempts a `conda-env` when you use `virsorter run`, and then it crashes - did you need to do anything else to stop `conda` getting upset?

**Comment by jiarong** [comment]
@all, I just had chance to test this. It looks like the new mamba version 2 is the issue. You can change to mamba version 1 to resolve this issue.
```
# get into the vs2 env
conda activate vs2
# downgrade mamba to v1
mamba installl -c conda-forge mamba=1
```

**Comment by anniedussault** [comment]
thanks @jiarong thanks for the solution! I was able to install your app and run it!

**Comment by sohini5981** [comment]
Hello @jiarong 

I have installed vs2 using mamba (ver  1.5.9)
base                  *  /storage/home/svg5863/miniforge3
vs2                      /storage/home/svg5863/miniforge3/envs/vs2. However, the problem @joelive faced persists for me. When I try to setup the db folder, the job exists with the following errors:


/storage/home/svg5863/miniforge3/envs/vs2/bin/virsorter setup -d db -j 4
[2024-11-22 13:05 INFO] Setting up VirSorter2 database; this might take ~10 mins and only needs to be done once.
[2024-11-22 13:05 INFO] saving /storage/home/svg5863/db as DBDIR to config file /storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml
CreateCondaEnvironmentException:
Could not create conda environment from /storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/site-packages/virsorter/rules/../envs/vs2.yaml:
/storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/argparse.py:1930: FutureWarning: `remote_definition` is deprecated and will be removed in 25.9. Use `conda env create --file=URL` instead.
  action(self, namespace, argument_values, option_string)
Collecting package metadata (repodata.json): ...working... 
  File "/storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/site-packages/snakemake/deployment/conda.py", line 350, in create
[2024-11-22 13:06 ERROR] CreateCondaEnvironmentException:
Could not create conda environment from /storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/site-packages/virsorter/rules/../envs/vs2.yaml:
/storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/argparse.py:1930: FutureWarning: `remote_definition` is deprecated and will be removed in 25.9. Use `conda env create --file=URL` instead.
  action(self, namespace, argument_values, option_string)
Collecting package metadata (repodata.json): ...working... 
  File "/storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/site-packages/snakemake/deployment/conda.py", line 350, in create
[2024-11-22 13:06 INFO] First attempt failed; trying the second time.
CreateCondaEnvironmentException:
Could not create conda environment from /storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/site-packages/virsorter/rules/../envs/vs2.yaml:
/storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/argparse.py:1930: FutureWarning: `remote_definition` is deprecated and will be removed in 25.9. Use `conda env create --file=URL` instead.
  action(self, namespace, argument_values, option_string)
Collecting package metadata (repodata.json): ...working... 
  File "/storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/site-packages/snakemake/deployment/conda.py", line 350, in create
[2024-11-22 13:07 CRITICAL] Command 'snakemake --snakefile /storage/home/svg5863/miniforge3/envs/vs2/lib/python3.10/site-packages/virsorter/rules/setup-retry.smk --directory /storage/home/svg5863/db --quiet --config Skip_deps_install=False --jobs 4 --rerun-incomplete --latency-wait 600 --nolock  --use-conda --conda-prefix /storage/home/svg5863/db/conda_envs --conda-frontend mamba ' returned non-zero exit status 1.
 Any advice??

**Comment by anniedussault** [comment]
@sohini5981 Have you tried the solution here ? https://github.com/jiarong/VirSorter2/issues/222#issuecomment-2452806948 

I had a smilar issue to what you're describing and this resolved my issue. I would start with that!

**Comment by jiarong** [comment]
> @sohini5981 Have you tried the solution here ? [#222 (comment)](https://github.com/jiarong/VirSorter2/issues/222#issuecomment-2452806948)
> 
> I had a smilar issue to what you're describing and this resolved my issue. I would start with that!

Right, it's the same issue. The trick is to change mamba within the VS2 env. The problem is the mamba within the VS2 env, not the mamba in your base env.

**Comment by sohini5981** [comment]
Thank you all for addressing the issue. I did change the mamba ver within
the vs2 env. It did not help. Hence, I used the aptainer  method to install
the whole thing. That worked and hopefully now I will be able to run it.
Thank you all.

**Comment by jiarong** [comment]
@all, I  have fixed this issue in bioconda version. Just upgarde and you should be able to run w/o manually downgrading mamba to v1.

**Comment by bioinfogini** [comment]
@jiarong thanks for posting the solution about downgrading mamba when facing libmamba Non-conda folder exists at prefix Error! It worked also for issues not related to VirStorter2, but to Metagenome atlas as well!

**Comment by SilasK** [comment]
Thank you for the fix. any idea hopw this could be permanently solved. libmamba v1 is no longer maintained and eventually we should swicht to v2..

**Comment by jiarong** [comment]
My understanding is that the issue is caused by the incompatibility between snakemake and mamba2. We can switch back to conda since it's conda's resolver is pretty decent now, or just wait till snakemake resolve the issue with mamba2.

**Comment by mrqasimasif** [comment]
> mamba installl -c conda-forge mamba=1

I tried multiple things but the mamba downgrade solution worked for me.
